### DISCUSSION TITLE
Jotai with React 19 (RSC, Server Action)?

### DISCUSSION THREAD

**Discussion by himself65** [pmndrs/jotai#2429] [General]
I've been thinking about this for a few days. The ideas might be immature, but I should share them with you guys and discuss them.

# atom in a server component (RSC)?

This works better than I expected. Imaging you have `userIdAtom` and, `userAtom,` `postAtom` in the server component; you can just set `userId` and then get `userAtom` and `postAtom` just in one call because of dependency tracking.

```tsx
"use server"
// put these atom into a single file, logic will be much cleaner
const userIdAtom = atom<string | null>(null)
const userAtom atom(async () => {
	return sql`SELECT ...`
})
const postAtom = atom(async () => {
	return sql`SELECT ...`
})


// imaging there is a such API in the future React framework
export const beforeRenderApp = () => {
	const store = createStore()
	const { userId } = parse(cookie().get('session'))
	store.set(userIdAtom, userId)
	return {
		store
	}
}

export const afterRenderApp = (ctx) => {
	// cleanup atom
	ctx.store.set(userIdAtom, null)
}

export const App = async () => {
	// it's clean code
	const { store } =getContext()
	const user = await store.get(userAtom)
	const post = await store.get(postAtom)
	return (
		<div>
			Hello! ${user.name}
			<ul>
				{...}
			</ul>
		</div>
	)
}
```

You can do this in next.js or waku right now, but you might need add some wrapper code to make each request have a new store.

# atom + server action?

server action is a good API for making interactive UI between client side and server side.

for example, [vercel rsc](https://sdk.vercel.ai/docs/concepts/ai-rsc) give us a way to make AI bot easily with Generative UI, which you can let AI to response UI directly also combine the functionality using server action.

So I'm thinking if we can also let atoms directly pass to the client side, for example
```ts
// #file: server-atom.ts
import { userIdAtom } from "client-atom"
export const postAtom = atom(async (get) => {
	"use server"
	const userId = get(userIdAtom)
	if (!userId) {
		throw new Error("NO USER")
	}
	return sql`...`
})
// #file: client-atom.ts
export const userIdAtom = atom<string | null>(null)
```

There is some restriction like in server side there cannot have primitiveAtom (no state in server!)

## WHY?

because atoms are built top of object reference.

If you read react source code, you will find that server action is actually a object reference and pass to the client side, and client side will make a POST request to call that reference. But restriction is that server action can be only pass through `props` inside a React Component. Nowadays people usually use `createContext` to passthrough the server actions. If we can by someway to allow server/client atoms. This would be very useful in many cases.

But this must violate the **standard** react (even I don't know what is a standard for RSC, feel like next.js, waku, parcel treat the same code differently), and this would also need we developed one more compiler (for now we call `jotai-rsc`). For the above code, it will be transcompile to:

```tsx
const userIdAtom = createAtomReference('client-atom.ts#userIdAtom')

export const postAtom = atom(async (get) => {
	"use server"
	const userId = get(userIdAtom)
	if (!userId) {
		throw new Error("NO USER")
	}
	return sql`...`
})

const App = async () => {
	const post = await store.get(postAtom)	// <- if you call `store` API inside UI, compiler will collect this atom for rendering.
	return ...
}
```

and for this request call, it would be like

```:
req('/index.html', formData={
	'client-atom.ts#userIdAtom': currentUserId
})
```

```tsx
// #file: server-atom.ts
const userIdAtom = createAtomReference('client-atom.ts#userIdAtom')

export const switchUserAtom = atom(null, (get, set, newUserId) => {
	if (validate(newUserId)) {
		set(userIdAtom, newUserId)
	}
	throw new Error('')
})
// #file: client-atom.ts
const switchUserAtom = createAtomReference("server-atom.ts#switchUserAtom")

export const userIdAtom = atom<string | null>(null)

const Login = () => {
	return (
		<button onClick={() => store.set(switchUserAtom, "userNamefromInput")}>
			Switch User
		</button>
	)
}
```

and for this server action call, it would be like:
```
req('/index', method='POST', formData={
	actionId: 'server-atom.ts#switchUserAtom',
	params: ['userNamefromInput']
})
->
res(json={
	'client-atom.ts#userIdAtom': 'newUserId'
})
```

Furthermore, this could be framework-agnostic; you need two extra options, `on('atomUpdate')` callback and `send(data: Data)` util, and it should also work in a simple HTTP server, RPC server, and non-React server.

## Workaround for now

There is some workaround to make this happen, but the code will be very tricky and ugly. you will need `Provider` + `createContext.` and make the atom await until UI is rendered using `React.use(promise).` I will not show the code here for now

**Comment by dai-shi** [comment]
# atom in a server component (RSC)?

> you might need add some wrapper code to make each request have a new store.

Yeah, that's the point.

# atom + server action?

> if we can also let atoms directly pass to the client side

I had a different example: https://github.com/dai-shi/waku/pull/11

Yeah, I'm looking forward to exploring this once we reach Waku v1-alpha.

> this could be framework-agnostic

I hadn't thought about it. Interesting.
